File: widgets/SelectUnits/SelectUnits.js
import React, { Component } from 'react';
import { PythonInterface } from 'reactopya';
import UnitSelector from '../jscommon/UnitSelector.js';
const config = require('./SelectUnits.json');

export default class SelectUnits extends Component {
    static title = 'Select unit ids'
    static reactopyaConfig = config
    constructor(props) {
        super(props);
        this.state = {
            // javascript state
            sorting: null,
            
            // python state
            unit_ids: [],
            status: '',
            status_message: '',

            //
            selectedUnits: {}
        }
    }
    componentDidMount() {
        this.pythonInterface = new PythonInterface(this, config);
        this.pythonInterface.start();
        this.pythonInterface.setState({
            sorting: this.props.sorting
        });
    }
    componentWillUnmount() {
        this.pythonInterface.stop();
    }
    _handleChange = (selectedUnits) => {
        this.setState({
            selectedUnits: selectedUnits
        });
        this.props.onSelectedUnitsChanged && this.props.onSelectedUnitsChanged(selectedUnits);
    }
    render() {
        let style0 = {
            position: 'relative',
            left: 0,
            right: 0,
            top: 0,
            height: this.props.height || 200,
            background: 'lightgray',
            overflow: 'auto'
        };
        return (
            <RespectStatus {...this.state}>
                <div style={style0}>
                    <UnitSelector
                        all_unit_ids={this.state.unit_ids}
                        selectedUnits={this.state.selectedUnits}
                        onChange={this._handleChange}
                    />
                </div>
            </RespectStatus>
        )
    }
}

class RespectStatus extends Component {
    state = {}
    render() {
        switch (this.props.status) {
            case 'running':
                return <div>Running: {this.props.status_message}</div>
            case 'error':
                return <div>Error: {this.props.status_message}</div>
            case 'finished':
                return this.props.children;
            default:
                return <div>Unknown status: {this.props.status}</div>
        }
    }
}

File: widgets/SelectUnits/SelectUnits.py
import traceback
from ..pycommon.autoextractors import AutoSortingExtractor


class SelectUnits:
    def __init__(self):
        super().__init__()

    def javascript_state_changed(self, prev_state, state):
        self._set_status('running', 'Running SelectUnits')

        sorting0 = state.get('sorting', None)
        if not sorting0:
            self._set_error('Missing: sorting')
            return
        try:
            self._sorting = AutoSortingExtractor(sorting0)
        except Exception as err:
            traceback.print_exc()
            self._set_error('Problem initiating sorting: {}'.format(err))
            return
        
        self._set_state(
            unit_ids=self._sorting.get_unit_ids(),
            status='finished',
            status_message='finished'
        )

    def _set_state(self, **kwargs):
        self.set_state(kwargs)
    
    def _set_error(self, error_message):
        self._set_status('error', error_message)
    
    def _set_status(self, status, status_message=''):
        self._set_state(status=status, status_message=status_message)